File: contrib/trackio-rs/src/batcher.rs
use crate::client::Client;
use crate::log_item::LogItem;
use crate::payload::BulkLogPayload;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

pub struct Batcher {
    client: Arc<Client>,
    buf: Arc<Mutex<Vec<LogItem>>>,
    max_batch: usize,
    flush_interval: Duration,
}

impl Batcher {
    pub fn new(client: Arc<Client>, max_batch: usize, flush_interval: Duration) -> Self {
        let b = Self {
            client: client.clone(),
            buf: Arc::new(Mutex::new(Vec::new())),
            max_batch,
            flush_interval,
        };

        // background auto-flush thread
        {
            let client_ref = client.clone();
            let buf_ref = b.buf.clone();
            let interval = b.flush_interval;
            thread::spawn(move || loop {
                thread::sleep(interval);
                let _ = Batcher::flush_static(&client_ref, &buf_ref);
            });
        }

        b
    }

    pub fn enqueue(&self, item: LogItem) {
        let mut buf = self.buf.lock().unwrap();
        buf.push(item);
        if buf.len() >= self.max_batch {
            let _ = Batcher::flush_static(&self.client, &self.buf);
        }
    }

    pub fn flush(&self) -> anyhow::Result<()> {
        Self::flush_static(&self.client, &self.buf)
    }

    fn flush_static(client: &Arc<Client>, buf: &Arc<Mutex<Vec<LogItem>>>) -> anyhow::Result<()> {
        let items = {
            let mut guard = buf.lock().unwrap();
            if guard.is_empty() {
                return Ok(());
            }
            let items = guard.clone();
            guard.clear();
            items
        };

        let metrics_list: Vec<_> = items.iter().map(|it| it.metrics.clone()).collect();
        let steps: Vec<_> = items.iter().map(|it| it.step.unwrap_or(-1)).collect();
        let timestamps: Vec<_> = items.iter().map(|it| it.timestamp.clone()).collect();

        let payload = BulkLogPayload {
            project: client.project.clone(),
            run: client.run.clone(),
            metrics_list,
            steps,
            timestamps,
            config: None,
        };

        client.post_json("/api/bulk_log", &payload)?;
        Ok(())
    }
}

File: contrib/trackio-go/batcher.go
package trackio

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"sync"
	"time"
)

type batcher struct {
	c        *Client
	mu       sync.Mutex
	buf      []LogItem
	maxBatch int
	ticker   *time.Ticker
	stopped  chan struct{}
}

func newBatcher(c *Client) *batcher {
	max := envInt("TRACKIO_MAX_BATCH", 128)
	interval := time.Duration(envInt("TRACKIO_FLUSH_INTERVAL_MS", 200)) * time.Millisecond
	b := &batcher{
		c:        c,
		maxBatch: max,
		ticker:   time.NewTicker(interval),
		stopped:  make(chan struct{}),
	}
	go b.loop()
	return b
}

func (b *batcher) loop() {
	for {
		select {
		case <-b.ticker.C:
			_ = b.flush(context.Background())
		case <-b.stopped:
			return
		}
	}
}

func (b *batcher) enqueue(it LogItem) {
	b.mu.Lock()
	b.buf = append(b.buf, it)
	shouldFlush := len(b.buf) >= b.maxBatch
	b.mu.Unlock()
	if shouldFlush {
		_ = b.flush(context.Background())
	}
}

func (b *batcher) flush(ctx context.Context) error {
	b.mu.Lock()
	items := b.buf
	b.buf = nil
	b.mu.Unlock()

	if len(items) == 0 {
		return nil
	}

	metricsList := make([]map[string]any, 0, len(items))
	steps := make([]int, 0, len(items))
	timestamps := make([]string, 0, len(items))

	for _, it := range items {
		if it.Metrics == nil {
			it.Metrics = map[string]any{}
		}
		metricsList = append(metricsList, it.Metrics)

		if it.Step != nil {
			steps = append(steps, *it.Step)
		} else {
			// keep step vector aligned; use -1 as "unset"
			steps = append(steps, -1)
		}

		if it.Timestamp == "" {
			timestamps = append(timestamps, "")
		} else {
			timestamps = append(timestamps, it.Timestamp)
		}
	}

	payload := bulkLogPayload{
		Project:     b.c.project,
		Run:         b.c.run,
		MetricsList: metricsList,
		Steps:       steps,
		Timestamps:  timestamps,
		// Config:    nil, // set if you want to send config once
	}

	// Try modern REST route first, then legacy gradio route
	if err := b.c.tryPost(ctx, "/api/bulk_log", payload); err == nil {
		return nil
	}
	if err := b.c.tryPost(ctx, "/gradio_api/bulk_log", payload); err == nil {
		return nil
	}
	return fmt.Errorf("trackio: unable to POST to either /api/bulk_log or /gradio_api/bulk_log")
}

func envInt(k string, def int) int {
	if v := os.Getenv(k); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			return n
		}
	}
	return def
}
